### PR TITLE
My final features

### PR DESCRIPTION
**Pull request by ricdezen**
## Premise

I hadded so many things I lost count. But there is nothing more I'm willing to do tbh.

## Changes

- Added Contact selection
- Added Swipe gesture to the log, left deletes, right adds to contacts if not already present.
- Fixed requests and responses
- Added Foreground service
- Added button in the Alarm Activity to shut down said service.
- Added answering the ring request with how much time elapsed.
- Added a timeout to the rings.
- Plugged in log 100%.
- Plugged in encryption for the sending.
- Added dialog inviting user to set a password.
- Various fixes to navigation.

Probably something else.
I have asserted everything works. I invite you to do so at least once yourself before Friday.

### CODE REVIEW

File: app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+ *
+ * @author Riccardo De Zen.
+ */
+public class ResponseManager {
+
+    private static ResponseManager instance;
+
+    private LocationManager locationManager = new LocationManager();
+    private MessageParser messageParser = new MessageParser();
+    private SMSContactManager contacts;
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java:37]: Codacy Issue found: [Avoid unused private fields such as 'contacts'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

**Comment by ricdezen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 6
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java  2
- app/src/main/java/ingsw/group1/findmyphone/ReceivedMessageManager.java  3
- app/src/main/java/ingsw/group1/findmyphone/MessageParser.java  9
- app/src/main/java/ingsw/group1/findmyphone/activity/AlarmActivity.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/items/LogMap.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/CreateContactFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/RingService.java  5
- app/src/main/java/ingsw/group1/findmyphone/SharedData.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/SettingsFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogItemSwipeCallback.java  3
- app/src/main/java/ingsw/group1/findmyphone/InfoDialog.java  4
- app/src/main/java/ingsw/group1/findmyphone/location/LocationResponseCommand.java  1
- app/src/main/java/ingsw/group1/findmyphone/RequestManager.java  2
- app/src/main/java/ingsw/group1/findmyphone/MessageType.java  1
- app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java  5
- app/src/test/java/ingsw/group1/findmyphone/MessageParserTest.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829&bid=16815789)

File: app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java (hunk 1 above)
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java:37]: Codacy Issue found: [Perhaps 'contacts' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

File: app/src/main/java/ingsw/group1/findmyphone/RingService.java [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+     * Method to start a 2 minute timeout after which the service is shut down.
+     */
+    private void startTimeout() {
+        final int timeoutTime = 120000;
+        new CountDownTimer(timeoutTime, 1000) {
+            /**
+             * No action is performed on tick
+             */
+            @Override
+            public void onTick(long l) {
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/RingService.java:180]: Codacy Issue found: [Document empty method body](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

File: app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+     * - {@link MessageType#RING_RESPONSE} The Ring Event is recorded in the Log Database.
+     * - {@link MessageType#LOCATION_REQUEST} The asynchronous location retrieval is started and
+     * will send back the result when available.
+     * - {@link MessageType#LOCATION_RESPONSE} The Location Event is recorded in the Log Database.
+     *
+     * @param callingContext The calling context, used if starting a Service or such.
+     * @param message        The message to handle.
+     */
+    public void performActionBasedOnMessage(Context callingContext, @NonNull SMSMessage message) {
+        switch (messageParser.getMessageType(message)) {
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/ResponseManager.java:75]: Codacy Issue found: [Switch statements should have a default label](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

File: app/src/main/java/ingsw/group1/findmyphone/log/LogItemSwipeCallback.java [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+
+    /**
+     * When the item is swiped left, it is removed from the adapter.
+     *
+     * @param viewHolder The ViewHolder that triggered the swipe.
+     * @param direction  The direction in which the swipe happened.
+     */
+    @Override
+    public void onSwiped(@NonNull RecyclerView.ViewHolder viewHolder, int direction) {
+        switch (direction) {
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/log/LogItemSwipeCallback.java:86]: Codacy Issue found: [Switch statements should have a default label](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

File: app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java [hunk 5]
```diff
@@ -155,6 +216,14 @@ public boolean onCreateOptionsMenu(Menu menu) {
     @Override
     public boolean onOptionsItemSelected(@NonNull MenuItem item) {
         NavController navController = Navigation.findNavController(this, R.id.nav_host_fragment);
+        switch (item.getItemId()) {
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java:219]: Codacy Issue found: [Switch statements should have a default label](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5114829)

**Review by bortoletti-giorgia — APPROVED** [review]
I have tested everything and it seems to me that there is nothing to report.
I had also implemented the contact selection in this [pull request](https://github.com/RicDeZen/ingsw-group1-findmyphone/pull/44) but we keep this that is already hooked to the rest.
Good job. I definitely approve.

**Pull request merged**